lib/matplotlib/legend.py

```diff
@@ -409,42 +409,27 @@ def _set_loc(self, loc):
         # _legend_box will draw itself at the location of the return
         # value of the find_offset.
         self._loc_real = loc
-        if loc == 0:
-            _findoffset = self._findoffset_best
-        else:
-            _findoffset = self._findoffset_loc
-
-#       def findoffset(width, height, xdescent, ydescent):
-#           return _findoffset(width, height, xdescent, ydescent, renderer)
-
-        self._legend_box.set_offset(_findoffset)
-
-        self._loc_real = loc
         self.stale = True
 
     def _get_loc(self):
         return self._loc_real
 
     _loc = property(_get_loc, _set_loc)
 
-    def _findoffset_best(self, width, height, xdescent, ydescent, renderer):
-        "Helper function to locate the legend at its best position"
-        ox, oy = self._find_best_position(width, height, renderer)
-        return ox + xdescent, oy + ydescent
-
-    def _findoffset_loc(self, width, height, xdescent, ydescent, renderer):
-        "Helper function to locate the legend using the location code"
+    def _findoffset(self, width, height, xdescent, ydescent, renderer):
+        "Helper function to locate the legend"
 
-        if iterable(self._loc) and len(self._loc) == 2:
-            # when loc is a tuple of axes(or figure) coordinates.
-            fx, fy = self._loc
-            bbox = self.get_bbox_to_anchor()
-            x, y = bbox.x0 + bbox.width * fx, bbox.y0 + bbox.height * fy
-        else:
+        if self._loc == 0:  # "best".
+            x, y = self._find_best_position(width, height, renderer)
+        elif self._loc in Legend.codes.values():  # Fixed location.
             bbox = Bbox.from_bounds(0, 0, width, height)
             x, y = self._get_anchored_bbox(self._loc, bbox,
                                            self.get_bbox_to_anchor(),
                                            renderer)
+        else:  # Axes or figure coordinates.
```

Comment:
The condition effectively is the same as before.

Previous impl:
- if loc == 0, use findoffset_best, which directly calls find_best_position.
- otherwise, use findoffset_loc:
  - if len(loc) == 2, use fixed position
  - otherwise, use get_anchored_bbox (fixed location, assumed to be in Legend.codes.values)

New impl:
- if loc == 0, use find_best_position
- if fixed location, in Legend.codes.values, use get_anchored_bbox
- otherwise, assume fixed position

It is correct that the error on invalid locations would be different though.  The previous implementation would raise an AssertionError (`assert loc in range(1, 11)  # called only internally` in get_anchored_bbox), the current one raises a TypeError (when trying to iterate over self._loc).